tests/src/python/test_selective_masking.py

```diff
@@ -1329,6 +1330,100 @@ def test_vector_random_generator_fill(self):
 
         self.check_layout_export("layout_export_random_generator_fill", 0, [layer], extent=extent)
 
+    def test_layer_opacity(self):
+        """
+        Test that layer opacity is respected during exports
+        """
+        self.assertTrue(QgsProject.instance().read(os.path.join(unitTestDataPath(), "selective_masking.qgs")))
+
+        polys = QgsProject.instance().mapLayersByName('polys')[0]
+        self.assertTrue(polys)
+        simple_fill = QgsFillSymbol.createSimple({'color': '#ffaaaa', 'style_border': 'no'})
+        polys.setRenderer(QgsSingleSymbolRenderer(simple_fill))
+        polys.setOpacity(0.5)
+
+        points = QgsProject.instance().mapLayersByName('points')[0]
+        self.assertTrue(points)
+
+        simple_marker = QgsMarkerSymbol.createSimple({'color': '#0000ff', 'size': '30', 'outline_style': 'no'})
+        points.setRenderer(QgsSingleSymbolRenderer(simple_marker))
+
+        self.assertTrue(len(polys.labeling().subProviders()), 1)
+        settings = polys.labeling().settings()
+        fmt = settings.format()
+        fmt.setFont(QgsFontUtils.getStandardTestFont("Bold"))
+        fmt.setSize(32)
+        fmt.setSizeUnit(QgsUnitTypes.RenderPoints)
+
+        mask = fmt.mask()
+        mask.setEnabled(True)
+        mask.setSize(9)
+        mask.setSizeUnit(Qgis.RenderUnit.Millimeters)
+        mask.setMaskedSymbolLayers([QgsSymbolLayerReference(points.id(), simple_marker[0].id())])
+
+        fmt.setMask(mask)
+
+        settings.setFormat(fmt)
+        polys.labeling().setSettings(settings)
+
+        map_settings = QgsMapSettings()
+        crs = polys.crs()
+        map_settings.setExtent(polys.extent())
+        map_settings.setBackgroundColor(QColor(152, 219, 249))
+        map_settings.setOutputSize(QSize(800, 600))
+        map_settings.setOutputDpi(72)
+        map_settings.setFlag(QgsMapSettings.Antialiasing, True)
+        map_settings.setFlag(QgsMapSettings.ForceVectorOutput, True)
+        map_settings.setFlag(QgsMapSettings.UseAdvancedEffects, True)
+        map_settings.setDestinationCrs(crs)
+        map_settings.setLayers([points, polys])
+
+        im = QImage(map_settings.outputSize(), QImage.Format_RGB32)
+        im.setDotsPerMeterX(int(map_settings.outputDpi() / 25.4 * 1000))
+        im.setDotsPerMeterY(int(map_settings.outputDpi() / 25.4 * 1000))
+        im.fill(Qt.transparent)
+        p = QPainter(im)
+
+        job = QgsMapRendererCustomPainterJob(map_settings, p)
+        job.start()
+        job.waitForFinished()
+
+        p.end()
+
+        self.assertTrue(self.imageCheck('layer_opacity', 'layer_opacity', im))
+
+        polys.setOpacity(1.0)
```

Comment:
In the resulting expected [image](https://github.com/qgis/QGIS/pull/51995/files#diff-0b8924dfd96c55d5ee160cf1543fcf78059a04631b3b004544717515c6d5b00a) points are not masked anymore, although they should be no?